core/shared/src/main/scala/zio/VersionSpecific.scala

```diff
@@ -72,25 +79,25 @@ private[zio] trait VersionSpecific {
    * `Tag[A with B]` should produce `Set(Tag[A], Tag[B])`
    */
   private[zio] def taggedGetServices[A](t: LightTypeTag): Set[LightTypeTag] =
-    taggedServices.computeIfAbsent(t, taggedServicesFn)
-
-  private val taggedSubtypes: JMap[(LightTypeTag, LightTypeTag), BoxedBool] =
-    Platform.newConcurrentMap()(Unsafe.unsafe)
-
-  private val taggedServices: JMap[LightTypeTag, Set[LightTypeTag]] =
-    Platform.newConcurrentMap()(Unsafe.unsafe)
-
-  private[this] val taggedIsSubtypeFn =
-    new java.util.function.Function[(LightTypeTag, LightTypeTag), BoxedBool] {
-      override def apply(tags: (LightTypeTag, LightTypeTag)): BoxedBool =
-        if (tags._1 <:< tags._2) BoxedBool.True else BoxedBool.False
+    // NOTE: See `taggedIsSubtype` for implementation notes
+    taggedServices.get(t) match {
+      case null =>
+        val v = t.decompose
+        taggedServices.putIfAbsent(t, v)
+        v
+      case v => v
     }
 
-  private[this] val taggedServicesFn =
-    new java.util.function.Function[LightTypeTag, Set[LightTypeTag]] {
-      override def apply(tag: LightTypeTag): Set[LightTypeTag] =
-        tag.decompose
-    }
+  private val taggedSubtypes: ConcurrentHashMap[(LightTypeTag, LightTypeTag), java.lang.Boolean] = {
+    /*
+     * '''NOTE''': Larger maps have lower chance of collision which offers better
+     * read performance and smaller chance of entering synchronized blocks during writes
+     */
+    new ConcurrentHashMap[(LightTypeTag, LightTypeTag), java.lang.Boolean](1024)
```

Comment:
Any bench improvement? I wonder if we could play with the load factor. Google hash maps use 0.5 overall.

Reply:
> Any bench improvement

It's a bit hard to quantify because of internal caching in the `ZEnvironment`, but there is ~10% throughput improvement when accessing the Scope from multiple threads. Although I think the improvement is likely to be larger in applications that run on environments with large number of CPUs and that rely on ZEnvironment heavily as I suspect there will be more contention there